### ISSUE TITLE
The Lottery Ticket Hypothesis: Finding Sparse, Trainable Neural Networks

### ISSUE THREAD

**Issue by reyllama** [reyllama/paper-reviews#46]
## Motivation

- Neural network pruning can reduce parameter counts to up to 90% without sacrificing the performance.
- The pruned, sparse architecture is difficult to train from the start, which would've enabled saving time as well as memory.

## Key Idea

- There are subnetworks whose initialization makes them capable of training effectively, thus finding them could reduce training memory and time greatly.
- Iteratively prune smallest magnitude weights, and reset the remaining weights to their initialization.

## Method

1. Random initialize a neural network
2. Train the network for j iterations
3. Prune p% of the parameters by creating a mask 
4. Reset the remaining (unmasked) parameters to their initial values.

## Result

![image](https://user-images.githubusercontent.com/46308407/116640549-d13bfe80-a9a5-11eb-80da-91ea7faa9060.png)

- Until down to 21% of the parameters, the training speeds up (faster early stopping) and test accuracy matches or exceeds the full model.
- When reinitialize weights randomly in each iterative round, the test accuracy drops significantly.

![image](https://user-images.githubusercontent.com/46308407/116640646-08aaab00-a9a6-11eb-9576-50fd2ca727b7.png)

## Comment

- Surprising to see that how a subnetwork containing roughly 20% of the full parameters can give compatible results to the full model.
- Also astounding how initialization matters the performance of a model. The idea that over-parameterized models perform better because they afford more shots at lottery is intuitive and disappointing to a certain extent.
- The fact that over-parameterization assures modestly good convergence of gradient descent in non-convex optimization problem can be explained by this as well. Over-param models have many good starting points, for some of which the optimization problem is convex or close to convex.
